hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/AbstractDelegationTokenSecretManager.java

```diff
@@ -434,16 +446,17 @@ private void updateCurrentKey() throws IOException {
     DelegationKey newKey = new DelegationKey(newCurrentId, System
         .currentTimeMillis()
         + keyUpdateInterval + tokenMaxLifetime, generateSecret());
-    //Log must be invoked outside the lock on 'this'
-    logUpdateMasterKey(newKey);
     synchronized (this) {
+      storeDelegationKey(newKey);
```

Comment:
While I understand that we shouldn't update currentKey when some other thread is using it, but does the storeDelegationKey(newKey) also need to be inside the synchronized block?

Reply:
actually even updating currentKey needn't be in synchronized block because currentKey is volatile so the update will be visible. 

I had originally removed synchronized here but added it back to avoid the change in this PR and to make it seperately. 

wanted to keep the PR focused specifically on whats required for reliable stateless setup - in this particular case, trying to store in DB first before updating currentKey

Reply:
Won't updating currentKey need to be synchronized? Because if some other thread is using this in createPassword() and in between lines _identifier.setMasterKeyId(currentKey.getKeyId()) and createPassword(identifier.getBytes(), currentKey.getKey())_; the currentKey changes, won't that result in inconsistencies?